protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+- Distributed Record Store - A system to store provider records and relay records
+
+Each of these subsystem exposes a well known interface (see chapter 6 for Interfaces) and may use eachother in order to fulfil their goal. A global overview of the system is:
+
+```
+┌─────────────────────────────────────────────────────────────────────────────────┐
+│                                  libp2p                                         │
+└─────────────────────────────────────────────────────────────────────────────────┘
+┌─────────────────────────┐┌──────────────────────────┐┌──────────────────────────┐
+│       Peer Routing      ││      Swarm               ││ Distributed Record Store │
+└─────────────────────────┘└──────────────────────────┘└──────────────────────────┘
+```
+
+## 4.1 Peer Routing
+
+A Peer Routing subsystem, exposes an interface to identify which peers should a message be routed in the DHT. It receives a key and must return one or more PeerInfo objects.
+
+We present two examples of possible Peer Routing subsystems, the first based on a the Kademlia DHT and the second based on mDNS. Nevertheless, other Peer Routing mechanisms might be implemented, as long as they fulfil the same expectation and interface.
+
+### 4.1.1 kad-routing
+
+kad-routing implements the Kademlia Routing table, where each peer holds a set of k-buckets, each of them containing several PeerInfo from other peers in the network. In order to find the whereabouts of these peers, it implements 3 discovery mechanisms:
+
+- mDNS-discovery
+- random-walk
+- bootstrap-list
+
+```
+┌────────────────────────────────────────────────────────────────┐
+│       Peer Routing                                             │
+│┌──────────────────────────────────────────────────────────────┐│
+││ kad-routing ┌──────────────┐┌──────────────┐┌──────────────┐ ││
+││             │mDNS-discovery││random-walk   ││bootstrap-list│ ││
+││             └──────────────┘└──────────────┘└──────────────┘ ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ mDNS-routing                                                 ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ other-routing-mechanisms                                     ││
+│└──────────────────────────────────────────────────────────────┘│
+└────────────────────────────────────────────────────────────────┘
+```
+
+#### 4.1.1.1 mDNS-discovery
+
+mDNS-discovery uses mDNS to emit beacons in the local area network to find if there are more peers available, these peers are extremely interesting because our link to them will have an low latency, plus if we are disconnected from the network, we will be leverage the content on those peers have cached.
```

Comment:
- This whole thing should be under a heading of "Discovery Protocols"  
- This should also probably include how the DiscoveryProtocol interface is mapped to mDNS.
- maybe say:
> mDNS-discovery is a Discovery Protocol that uses mDNS (link to wikipedia) over local area networks. It emits mDNS beacons to find if there are more peers available. Local area network peers are very useful to peer-to-peer protocols, as low latency links are very useful.
> mDNS-discovery is a standalone protocol and does not depend on any other libp2p protocol. mDNS-discovery can yield peers available in the local area network, without relying on other infrastructure. This is particularly useful in intranets, networks disconnected from the internet backbone, and networks who temporarily loose links.
> mDNS-discovery can be configured per-service (i.e. discover only peers participating in a specific protocol, like IPFS), and with private networks (discover peers belonging to a private network).
> We are exploring ways to make mDNS-discovery beacons encrypted (so that other nodes in the local network cannot discern what service is being used). Though the nature of mDNS will always reveal local IP addresses.
> **Privacy Note**: mDNS advertises in local area networks, which reveals IP addresses to listeners in the same local network. It is not recommended to use this with privacy-sensitive applications or oblivious routing protocols.